nano/secure/ledger_set_confirmed.cpp

```diff
@@ -0,0 +1,161 @@
+#include <nano/secure/ledger.hpp>
+#include <nano/secure/ledger_set_confirmed.hpp>
+#include <nano/store/account.hpp>
+#include <nano/store/component.hpp>
+#include <nano/store/confirmation_height.hpp>
+#include <nano/store/pending.hpp>
+#include <nano/store/pruned.hpp>
+
+nano::ledger_set_confirmed::ledger_set_confirmed (nano::ledger const & ledger) :
+	ledger{ ledger }
+{
+}
+
+std::optional<nano::uint128_t> nano::ledger_set_confirmed::balance (store::transaction const & transaction, nano::account const & account_a) const
+{
+	auto block = get (transaction, head (transaction, account_a));
+	if (!block)
+	{
+		return std::nullopt;
+	}
+	return block->balance ().number ();
+}
+
+// Balance for account containing hash
+std::optional<nano::uint128_t> nano::ledger_set_confirmed::balance (store::transaction const & transaction, nano::block_hash const & hash) const
+{
+	if (hash.is_zero ())
+	{
+		return std::nullopt;
+	}
+	auto block = get (transaction, hash);
+	if (!block)
+	{
+		return std::nullopt;
+	}
+	return block->balance ().number ();
+}
+
+bool nano::ledger_set_confirmed::exists (store::transaction const & transaction, nano::block_hash const & hash) const
+{
+	auto block = ledger.store.block.get (transaction, hash);
+	if (!block)
+	{
+		return false;
+	}
+	auto info = ledger.store.confirmation_height.get (transaction, block->account ());
+	if (!info)
+	{
+		return false;
+	}
+	return block->sideband ().height <= info.value ().height;
+}
+
+bool nano::ledger_set_confirmed::exists_or_pruned (store::transaction const & transaction, nano::block_hash const & hash) const
+{
+	if (ledger.store.pruned.exists (transaction, hash))
+	{
+		return true;
+	}
+	auto block = ledger.store.block.get (transaction, hash);
+	if (!block)
+	{
+		return false;
+	}
+	auto info = ledger.store.confirmation_height.get (transaction, block->account ());
+	if (!info)
+	{
+		return false;
+	}
+	return block->sideband ().height <= info.value ().height;
+}
+
+std::optional<nano::account_info> nano::ledger_set_confirmed::get (store::transaction const & transaction, nano::account const & account) const
+{
+	auto info = ledger.store.confirmation_height.get (transaction, account);
+	if (!info)
+	{
+		return std::nullopt;
+	}
+	debug_assert (false);
```

Comment:
This doesn't look right. And the fact that no test fails with this assert means that we are missing an important test case or we have a bug (or it is one of the test we have disabled and still have not re-enabled).

Reply:
Fixed in:
https://github.com/nanocurrency/nano-node/pull/4486/commits/4c02ce53d17b0dca1a46b7fa3d945bfdf4018b59